Implement heuristics to drop out of Enhanced Security

ronan-apple opened the pull request:
#### 0c6fc3c778e57a5b66b2045881310b2a63068b0d
<pre>
Implement heuristics to drop out of Enhanced Security
<a href="https://bugs.webkit.org/show_bug.cgi?id=303390">https://bugs.webkit.org/show_bug.cgi?id=303390</a>
<a href="https://rdar.apple.com/165692583">rdar://165692583</a>

Reviewed by Matthew Finkel.

This change expands upon prior Enhanced Security adoption for HTTP which
now applies heuristics to determine when to drop out of Enhanced Security,
in particular, when we consider a site to have had meaningful prior
usage outside of Enhanced Security.

One requirement of this is to add a new WebsiteDataType for tracking when
sites have been seen outside of Enhanced Security, or only when Enhanced
Security was enabled. We persist this to a new db specifically for
Enhanced Security.

Additional tests have been implemented that check that these heuristics
apply successfully in conditions that we expect.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::monitoredDataTypes):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorageOrCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
(WebKit::toString):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setCurrentRequest):
(API::Navigation::setHasStorageForCurrentSite):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::hasStorageForCurrentSite const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _allWebsiteDataTypesIncludingPrivate]):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
(WebKit::enabledSitesMap):
(WebKit::didSitePreviouslyUseEnhancedSecurity):
(WebKit::trackSiteSeenOutsideEnhancedSecurity):
(WebKit::updateEnhancedSecurityDomains):
(WebKit::EnhancedSecurityTracking::initializeWithWebsiteDataStore):
(WebKit::EnhancedSecurityTracking::enableFor):
(WebKit::EnhancedSecurityTracking::trackChangingSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackSameSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackNavigation):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::hasLocalStorageOrCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::WebFramePolicyListenerProxy):
(WebKit::WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult):
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
(WebKit::WebFramePolicyListenerProxy::didReceiveInitialLinkDecorationFilteringData):
(WebKit::WebFramePolicyListenerProxy::didReceiveSiteHasStorageResults):
(WebKit::WebFramePolicyListenerProxy::use):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
(WebKit::WebFramePolicyListenerProxy::create):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setUpPolicyListenerProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForNewWindowAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::beginSiteHasStorageCheck):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultEnhancedSecurityDirectory):
(WebKit::WebsiteDataStore::enhancedSecuritySitesHolder):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp: Added.
(WebKit::EnhancedSecuritySitesHolder::sharedWorkQueueSingleton):
(WebKit::EnhancedSecuritySitesHolder::create):
(WebKit::EnhancedSecuritySitesHolder::EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::~EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesHolder::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesHolder::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesHolder::deleteSites):
(WebKit::EnhancedSecuritySitesHolder::deleteAllSites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h: Copied from Source/WebKit/UIProcess/EnhancedSecurityTracking.h.
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp: Added.
(WebKit::EnhancedSecuritySitesPersistence::EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::~EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::reportSQLError):
(WebKit::databasePath):
(WebKit::EnhancedSecuritySitesPersistence::checkedDatabase const):
(WebKit::EnhancedSecuritySitesPersistence::cachedStatement):
(WebKit::EnhancedSecuritySitesPersistence::openDatabase):
(WebKit::EnhancedSecuritySitesPersistence::deleteSite):
(WebKit::EnhancedSecuritySitesPersistence::deleteSites):
(WebKit::EnhancedSecuritySitesPersistence::deleteAllSites):
(WebKit::EnhancedSecuritySitesPersistence::enhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesPersistence::allEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesPersistence::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesPersistence::closeDatabase):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h: Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::hasLocalStorageOrCookies const):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::enhancedSecurityDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setEnhancedSecurityDirectory):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm:
(runHttpThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithLocalStorage):
(runHttpThenHttpsThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithCookiesThenHttps):
(runHttpThenNavigateToHttpsSetCookiesThenNavigateToHttpsAgain):
(runHttpRedirectsHttpsWithExplicitNavigationToMeaningfulSite):
(runWindowOpenThenNavigateToMeaningfulSite):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPI):
(enhancedSecuritySitesPath):
(emptyEnhancedSecuritySitesPath):
(createEnhancedSecuritySitesTable):
(addEnhancedSecuritySite):
(setUpEnhancedSecuritySeenValues):
(cleanUpEnhancedSecuritySites):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAndExpectations):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndNotSeenOutsideEnhancedSecurity):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndSeenOutsideEnhancedSecurity):

Canonical link: <a href="https://commits.webkit.org/303988@main">https://commits.webkit.org/303988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82a62312f938c7fb99327bbc987d05bf4f034339

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134212 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141796 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dd15edbd-51f0-49a9-86f2-7041d2aee882) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136082 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6587 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102639 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137159 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120294 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83432 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d323fc3b-7563-4b25-aba6-0bafaface733) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2577 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1608 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114124 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144439 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6393 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39009 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111010 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6480 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5383 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111255 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28218 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4805 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116568 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60164 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6445 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34786 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6292 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69911 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6533 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6399 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+    checkedDatabase()->disableThreadingChecks();
+
+    auto reportErrorAndCloseDatabase = [&](ASCIILiteral action) {
+        reportSQLError("openDatabase"_s, action);
```
Could use `__FUNCTION__` instead of manually typing the method names?

ronan-apple replied:
Ah, great point, thanks! I've used this throughout now.

sysrqb commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
```
Seems like a bad failure case. Maybe add a log, at least? Not sure what else we should do here.

ronan-apple replied:
Yep, good point, I've put in an error log here. This should only occur if `openDatabase` fails, in which case, I think any of the setup that we'd need to add/remove/query entries in the DB are likely to fail, I don't know that there's much else we could do here either. 

I suppose we'd hope that whatever failure occurred was intermittent and wouldn't occur on the next browser start, and we'd just fail to query/insert/remove entries in the interim?

sysrqb commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+
+    CheckedPtr insertSiteStatement = cachedStatement(StatementType::InsertSite).get();
+
+    int enhancedSecurityReason = static_cast<int>(reason);
```
nit: `uint8_t`

cdumez replied:
`enumToUnderlyingType()` is your friend.

ronan-apple replied:
Aha, perfect! Thanks, fixed.

sysrqb reviewed (commented):
Please add the new WebsiteDataType in `WebResourceLoadStatisticsStore::monitoredDataTypes`, too.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -35,6 +35,32 @@ namespace WebKit {
+
+void EnhancedSecurityTracking::initializeWithWebsiteDataStore(WebsiteDataStore& websiteDataStore)
+{
+    websiteDataStore.fetchEnhancedSecurityOnlyDomains([this](HashSet<RegistrableDomain>&& domains) {
```
I doubt safer cpp static analysis will be happy with this, assuming the lambda runs asynchronously. Capturing a raw pointer like this is hard to prove as safe. We likely want to capture a smart pointer.

ronan-apple replied:
Yep, very valid. I was looking at changing `EnhancedSecurityTracking` to support `WeakPtr`, but realised that we're only updating the static sites map in here, so it's actually perfectly feasible to change this to not capture `this` and instead call a static function in the completion handler.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -35,6 +35,32 @@ namespace WebKit {
+
+void EnhancedSecurityTracking::updateEnhancedSecurityDomains(HashSet<RegistrableDomain>&& domains)
+{
+    for (auto& domain : domains)
```
I think you can use `auto&&` here and `WTFMove(domain)` below.

ronan-apple replied:
I had a bit of trouble with this one - wondering if you have any ideas?

I was getting an error regarding const if I did as follows:

```
for (auto&& domain : domains)
    enabledSitesMap().add(WTFMove(domain), EnhancedSecurityReason::InsecureLoad
```

I think because the iterator returns a const key type? But this only failed on the Linux builds, not quite sure why that is?

Happy to change if you have an idea of what I'd need to do there - for now, I've changed it back to the copying that it was doing but with the additional clarity of `const auto&`.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -16425,6 +16434,14 @@ void WebPageProxy::waitForInitialLinkDecorationFilteringData(WebFramePolicyListe
 
+void WebPageProxy::beginSiteHasStorageCheck(const URL& url, API::Navigation& navigation, WebFramePolicyListenerProxy& listener)
+{
+    protectedWebsiteDataStore()->hasLocalStorageOrCookies(url, [weakThis = WeakPtr { *this }, navigation = Ref { navigation }, url = url.isolatedCopy(), listener = Ref { listener }] (bool siteHasStorage) mutable {
```
Why do we capture `weakThis` here, it doesn't appear to be used?

ronan-apple replied:
Great point! I think this was leftover from previous attempts at implementations, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -16425,6 +16434,14 @@ void WebPageProxy::waitForInitialLinkDecorationFilteringData(WebFramePolicyListe
 
+void WebPageProxy::beginSiteHasStorageCheck(const URL& url, API::Navigation& navigation, WebFramePolicyListenerProxy& listener)
+{
+    protectedWebsiteDataStore()->hasLocalStorageOrCookies(url, [weakThis = WeakPtr { *this }, navigation = Ref { navigation }, url = url.isolatedCopy(), listener = Ref { listener }] (bool siteHasStorage) mutable {
```
We should not need to isolatedCopy the url here? hasLocalStorageOrCookies() is NOT supposed to call the lambda on another thread, that would be error-prone.

ronan-apple replied:
Oh yep, great point, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -1077,4 +1085,59 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+
+    Vector<WebCore::RegistrableDomain> sites;
+
+    for (auto& origin : origins)
```
I think you can do something like:
```
auto sites = origins.map([](auto& origin) {
    return WebCore::RegistrableDomain { origin };
});
```

ronan-apple replied:
Aha, thank you! This works much nicer, yep.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+
+        auto enhancedSecuritySites = protectedThis->m_enhancedSecurityPersistence->fetchEnhancedSecurityOnlyDomains();
+
+        callOnMainRunLoop([enhancedSecuritySites = crossThreadCopy(enhancedSecuritySites), completionHandler = WTFMove(completionHandler)] mutable {
```
`crossThreadCopy(WTFMove(enhancedSecuritySites))` would be more efficient.

ronan-apple replied:
Ah ace, thank you, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+
+        auto enhancedSecuritySites = protectedThis->m_enhancedSecurityPersistence->fetchAllEnhancedSecuritySites();
+
+        callOnMainRunLoop([enhancedSecuritySites = crossThreadCopy(enhancedSecuritySites), completionHandler = WTFMove(completionHandler)] mutable {
```
`crossThreadCopy(WTFMove(enhancedSecuritySites))`

ronan-apple replied:
Thanks! Fixed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h:
```diff
@@ -0,0 +1,63 @@
+
+    EnhancedSecuritySitesHolder(const String& databasePath);
+
+    std::unique_ptr<EnhancedSecuritySitesPersistence> m_enhancedSecurityPersistence;
```
I recommend relying on `WTF_GUARDED_BY_CAPABILITY(sharedWorkQueueSingleton())` for this data member so that the compiler can check this is only used on the background queue (and never the main thread).

ronan-apple replied:
Ah nice, this is great, yep, implemented!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -1077,4 +1085,59 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+    for (auto& origin : origins)
+        sites.append(WebCore::RegistrableDomain(origin));
+
+    enhancedSecuritySitesHolder().deleteSites(sites, WTFMove(completionHandler));
```
Seems we can `WTFMove(sites)` here.

ronan-apple replied:
Ah yep, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+    });
+}
+
+void EnhancedSecuritySitesHolder::deleteSites(const Vector<WebCore::RegistrableDomain>& sites, CompletionHandler<void()>&& completionHandler)
```
`Vector<WebCore::RegistrableDomain>&&`

ronan-apple replied:
Ah yep, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+    if (sites.isEmpty())
+        return completionHandler();
+
+    sharedWorkQueueSingleton()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, sites = crossThreadCopy(sites), completionHandler = WTFMove(completionHandler)]() mutable {
```
`crossThreadCopy(WTFMove(sites))` would be more efficient.

ronan-apple replied:
Ah yes, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->m_enhancedSecurityPersistence->deleteSites(sites, WTFMove(wrappedCompletionHandler));
+        else
+            callOnMainRunLoop(WTFMove(completionHandler));
```
use-after-move of completionHandler no? I think you want `wrappedCompletionHandler();`.

ronan-apple replied:
Oh yes, good point! The other refactor removing the completion handler in deleteSites removed the need to have a wrapper, so this is fixed now from that change.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,157 @@
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->m_enhancedSecurityPersistence->deleteAllSites(WTFMove(wrappedCompletionHandler));
+        else
+            callOnMainRunLoop(WTFMove(completionHandler));
```
use after move?

ronan-apple replied:
Yep, good catch, thanks! As above, removed the need for `wrappedCompletionHandler` so this is now fixed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h:
```diff
@@ -0,0 +1,63 @@
+
+class EnhancedSecuritySitesPersistence;
+
+class EnhancedSecuritySitesHolder final : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<EnhancedSecuritySitesHolder, WTF::DestructionThread::Main> {
```
``WTF::DestructionThread::MainRunLoop` since we're in WebKit/. Especially important since this code runs in the UIProcess and the app may use both WK1 and WK2 on iOS and have a WebThread that is the main thread.

ronan-apple replied:
Oh I see! Okay, ace, thank you for the explanation! That's fixed now.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h:
```diff
@@ -0,0 +1,63 @@
+
+private:
+    // Shared WorkQueue is used to prevent race condition when delete and create Persistence for same database file.
+    static Ref<WorkQueue> sharedWorkQueueSingleton();
```
Should return a `WorkQueue&`. The `Singleton()` naming will make safer cpp static analysis happy without a smart pointer.

ronan-apple replied:
Nice, okay, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h:
```diff
@@ -0,0 +1,63 @@
+    // Shared WorkQueue is used to prevent race condition when delete and create Persistence for same database file.
+    static Ref<WorkQueue> sharedWorkQueueSingleton();
+
+    EnhancedSecuritySitesHolder(const String& databasePath);
```
`explicit EnhancedSecuritySitesHolder(const String& databasePath);`

ronan-apple replied:
Ah, yes, thank you, fixed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+        closeDatabase();
+}
+
+static constexpr auto sitesTableName = "sites"_s;
```
Can we move all these statics higher in the file, before the constructor/destructor. Seems weird in the middle here.

ronan-apple replied:
Yep, definitely, I've put them up to the top now.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+
+    const auto path = databasePath(directoryPath);
+
+    if (!checkedDatabase()->open(path, WebCore::SQLiteDatabase::OpenMode::ReadWriteCreate, WebCore::SQLiteDatabase::OpenOptions::CanSuspendWhileLocked))
```
We want a local variable for the database, calling `checkedDatabase()` is very inefficient as it keeps incrementing/decrementing a refcount.

ronan-apple replied:
Ah right, thank you - yes! I wasn't sure how to handle the CheckedPtr and the usage of `reportErrorAndCloseDatabase`, but I think I've handled it by clearing the local variable before closing the database (which destroys `m_sqliteDB` which the CheckedPtr is protecting).

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
```
Isn't it concerning we would leave data on disk when the database is not already open? I did not look much into it but thought I would point it out as it would be a bad privacy issue. We need to be able to delete data on disk, no matter our state.

ronan-apple replied:
This is a fair question, yeah.

I thought about this a bit more and I think this will only happen if we fail in the `openDatabase` call during initialisation. In that case, we won't have managed to add any data during this session, so the risk would be any older data. We could try to wipe the database entirely if something goes wrong in `openDatabase`, but that feels overkill for some of the error scenarios?

One other option is to store off a copy of the path, and,  in the `deleteSites` case where we want to remove all data, we could opt to try to wipe the directory if the DB wasn't open, as a bit of a fallback protection to still try to offer a means to remove the data?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+
+    auto deleteStatement = checkedDatabase()->prepareStatement(deleteAllSitesSQL);
+    if (!deleteStatement || !deleteStatement->executeCommand())
+        return reportSQLError("deleteAllSites"_s, "delete all sites"_s);
```
What calls the completion handler here?

ronan-apple replied:
Ah good point! This was addressed by removing the completion handler as the function was synchronous anyway.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+    for (auto& site : sites)
+        deleteSite(site);
+
+    completionHandler();
```
This function is synchronous . Why use a completion handler?

ronan-apple replied:
Great point - no need for it now, thanks! Fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,288 @@
+    completionHandler();
+}
+
+void EnhancedSecuritySitesPersistence::deleteAllSites(CompletionHandler<void()>&& completionHandler)
```
This function is synchronous . Why use a completion handler?

ronan-apple replied:
Great point - no need for it now, thanks! Fixed!

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h:
```diff
@@ -0,0 +1,74 @@
+class EnhancedSecuritySitesPersistence final {
+    WTF_DEPRECATED_MAKE_FAST_ALLOCATED(EnhancedSecuritySitesPersistence);
+public:
+    EnhancedSecuritySitesPersistence(const String&);
```
`explicit`

ronan-apple replied:
Ah, yep, fixed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h:
```diff
@@ -0,0 +1,74 @@
+    void deleteSites(const Vector<WebCore::RegistrableDomain>&, CompletionHandler<void()>&&);
+    void deleteAllSites(CompletionHandler<void()>&&);
+
+    HashSet<WebCore::RegistrableDomain> fetchEnhancedSecurityOnlyDomains();
```
I don't think we need `fetch()` in the name since this is synchronous.

ronan-apple replied:
Very valid, yep! Adjusted this and the other function below.

sysrqb commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt:
```diff
@@ -34,4 +34,4 @@ Registrable domain: localhost
     isPrevalentResource: Yes
     isVeryPrevalentResource: No
-    dataRecordsRemoved: 0
+    dataRecordsRemoved: 1
```
This shouldn't be related, and it's more likely a flakey test

ronan-apple replied:
Ah, so this was fairly consistent in the test failures across different EWS platforms. I could reproduce it locally too, but only when running a batch, when running just this single test it didn’t get the extra deleted record.

I didn’t quite fully understand the test - I think the extra record is likely our EnhancedSecurity new data type, as it only started triggering after the new monitoredDataTypes addition. 

If this only occurs when we’ve run other tests beforehand, I wonder if we’re getting a delayed effect from prior navigations saving the EnhancedSecurity record and then affecting this test?

ronan-apple replied:
We worked through this and figured out the reason for this. Rather than change this expectation, we realised we should be gating the tracking of the navigations (which then write the new EnhancedSecurity record) on the heuristics flag, so I've pushed a change to do this and reverted this expectation change.

sysrqb commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
```diff
@@ -39,6 +39,7 @@ WK_EXTERN NSString * const _WKWebsiteDataTypeAdClickAttributions WK_API_AVAILABL
 WK_EXTERN NSString * const _WKWebsiteDataTypePrivateClickMeasurements WK_API_AVAILABLE(macos(12.0), ios(15.0));
 WK_EXTERN NSString * const _WKWebsiteDataTypeAlternativeServices WK_API_AVAILABLE(macos(11.0), ios(14.0));
 WK_EXTERN NSString * const _WKWebsiteDataTypeFileSystem WK_API_DEPRECATED_WITH_REPLACEMENT("WKWebsiteDataTypeFileSystem", macos(13.0, 14.0), ios(16.0, 17.0));
+WK_EXTERN NSString * const _WKWebsiteDataTypeEnhancedSecuritySites WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
May need `visionos(WK_XROS_TBA)`, too?

ronan-apple replied:
Oh great question! Yes, this confused me - none of the others have visionos, but the fetchDataOfTypes and delete data types APIs _do_ seem exposed on visionOS….

ronan-apple replied:
Yes, I think you're right on this one, I think we do want the `visionos(WK_XROS_TBA)` here too - added.

sysrqb commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -16508,6 +16517,14 @@ void WebPageProxy::waitForInitialLinkDecorationFilteringData(WebFramePolicyListe
 
+void WebPageProxy::beginSiteHasStorageCheck(const URL& url, API::Navigation& navigation, WebFramePolicyListenerProxy& listener)
+{
+    protectedWebsiteDataStore()->hasLocalStorageOrCookies(url, [navigation = Ref { navigation }, url, listener = Ref { listener }] (bool siteHasStorage) mutable {
```
nit: boolean values like `siteHasStorage` should begin with a `is`/`has`/etc prefix, so maybe just `hasStorage`.

ronan-apple replied:
Ah, good point! Fixed.

sysrqb commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -170,8 +211,25 @@ void EnhancedSecurityTracking::trackNavigation(const API::Navigation& navigation
+        return;
+    }
+
+    bool siteHasStorage = navigation.hasStorageForCurrentSite();
```
minor nit: maybe `doesSiteHaveStorage`

ronan-apple replied:
Ah yep, fixed!

sysrqb commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -973,6 +973,8 @@ WebPageProxy::WebPageProxy(PageClient& pageClient, WebProcessProxy& process, Ref
     if (protectedPreferences()->siteIsolationEnabled())
         IPC::Connection::setShouldCrashOnMessageCheckFailure(true);
+
+    internals().enhancedSecurityTracker.initializeWithWebsiteDataStore(protectedWebsiteDataStore());
```
I think we can protect this with `if (protectedPreferences()->enhancedSecurityHeuristicsEnabled())`, too

ronan-apple replied:
Great point - yes! Fixed.

Commit:
Implement heuristics to drop out of Enhanced Security
https://bugs.webkit.org/show_bug.cgi?id=303390
rdar://165692583

Reviewed by Matthew Finkel.

This change expands upon prior Enhanced Security adoption for HTTP which
now applies heuristics to determine when to drop out of Enhanced Security,
in particular, when we consider a site to have had meaningful prior
usage outside of Enhanced Security.

One requirement of this is to add a new WebsiteDataType for tracking when
sites have been seen outside of Enhanced Security, or only when Enhanced
Security was enabled. We persist this to a new db specifically for
Enhanced Security.

Additional tests have been implemented that check that these heuristics
apply successfully in conditions that we expect.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::monitoredDataTypes):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorageOrCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
(WebKit::toString):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setCurrentRequest):
(API::Navigation::setHasStorageForCurrentSite):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::hasStorageForCurrentSite const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _allWebsiteDataTypesIncludingPrivate]):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
(WebKit::enabledSitesMap):
(WebKit::didSitePreviouslyUseEnhancedSecurity):
(WebKit::trackSiteSeenOutsideEnhancedSecurity):
(WebKit::updateEnhancedSecurityDomains):
(WebKit::EnhancedSecurityTracking::initializeWithWebsiteDataStore):
(WebKit::EnhancedSecurityTracking::enableFor):
(WebKit::EnhancedSecurityTracking::trackChangingSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackSameSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackNavigation):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::hasLocalStorageOrCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::WebFramePolicyListenerProxy):
(WebKit::WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult):
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
(WebKit::WebFramePolicyListenerProxy::didReceiveInitialLinkDecorationFilteringData):
(WebKit::WebFramePolicyListenerProxy::didReceiveSiteHasStorageResults):
(WebKit::WebFramePolicyListenerProxy::use):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
(WebKit::WebFramePolicyListenerProxy::create):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setUpPolicyListenerProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForNewWindowAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::beginSiteHasStorageCheck):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultEnhancedSecurityDirectory):
(WebKit::WebsiteDataStore::enhancedSecuritySitesHolder):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp: Added.
(WebKit::EnhancedSecuritySitesHolder::sharedWorkQueueSingleton):
(WebKit::EnhancedSecuritySitesHolder::create):
(WebKit::EnhancedSecuritySitesHolder::EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::~EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesHolder::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesHolder::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesHolder::deleteSites):
(WebKit::EnhancedSecuritySitesHolder::deleteAllSites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h: Copied from Source/WebKit/UIProcess/EnhancedSecurityTracking.h.
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp: Added.
(WebKit::EnhancedSecuritySitesPersistence::EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::~EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::reportSQLError):
(WebKit::databasePath):
(WebKit::EnhancedSecuritySitesPersistence::checkedDatabase const):
(WebKit::EnhancedSecuritySitesPersistence::cachedStatement):
(WebKit::EnhancedSecuritySitesPersistence::openDatabase):
(WebKit::EnhancedSecuritySitesPersistence::deleteSite):
(WebKit::EnhancedSecuritySitesPersistence::deleteSites):
(WebKit::EnhancedSecuritySitesPersistence::deleteAllSites):
(WebKit::EnhancedSecuritySitesPersistence::enhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesPersistence::allEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesPersistence::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesPersistence::closeDatabase):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h: Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::hasLocalStorageOrCookies const):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::enhancedSecurityDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setEnhancedSecurityDirectory):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm:
(runHttpThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithLocalStorage):
(runHttpThenHttpsThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithCookiesThenHttps):
(runHttpThenNavigateToHttpsSetCookiesThenNavigateToHttpsAgain):
(runHttpRedirectsHttpsWithExplicitNavigationToMeaningfulSite):
(runWindowOpenThenNavigateToMeaningfulSite):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPI):
(enhancedSecuritySitesPath):
(emptyEnhancedSecuritySitesPath):
(createEnhancedSecuritySitesTable):
(addEnhancedSecuritySite):
(setUpEnhancedSecuritySeenValues):
(cleanUpEnhancedSecuritySites):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAndExpectations):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndNotSeenOutsideEnhancedSecurity):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndSeenOutsideEnhancedSecurity):

Canonical link: https://commits.webkit.org/303988@main
